### PR TITLE
Da 2812 write dockerhub instrumentation in golang

### PR DESCRIPTION
**Pull request by enkhalifapro**

### CODE REVIEW

File: cmd/dads/dads.go [hunk 1]
```diff
@@ -17,6 +17,9 @@ func runDS(ctx *lib.Ctx) (err error) {
 		ds = &lib.DSJira{}
 	case lib.Groupsio:
 		ds = &lib.DSGroupsio{}
+	// todo: enable it later
```
Comment (lukaszgryglicki) [inline cmd/dads/dads.go:20]: Is this PR ready, or just in progress?

File: dockerhub/fetcher.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+// Params required parameters for dockerhub fetcher
+type Params struct {
+	Username       string
+	Password       string
+	BackendVersion string
+}
+
+// HttpClientProvider used in connecting to remote http server
+type HttpClientProvider interface {
+	Request(url string, method string, header map[string]string, body []byte) (statusCode int, resBody []byte, err error)
```
Comment (lukaszgryglicki) [inline dockerhub/fetcher.go:35]: You could you already existing Request - it supports everything here plus cookies, retrying, caching results) - up to you to decide.

File: utils/uuid/uuid.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package uuid
+
```
Comment (lukaszgryglicki) [inline utils/uuid/uuid.go:2]: Here, you *should* use the existing uuid generation support - it supports legacy-uuid mode which is very important. We need to generate exactly the same results as the python code does. Please at least take a look at uuid.go and see that in legacy mode it calls python uuid.py. If you want to use your own UUID code - make sure that results are always the same for python and go code. Also, note that there are two UUID generating methods - one for the unique id of raw documents and another one for affiliations/identity data. I know there should be just one, but we need exactly the same UUIDs generated - and code for those 2 situations was different in the original code.

File: utils/uuid/uuid_test.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+func TestGenerate(t *testing.T) {
+	// Arrange
+	args := []string{" abc ", "123"}
+
+	// Act
+	id, err := Generate(args...)
+
+	// Assert
+	assert.Equal(t, "18ecd81c8bb792b5c23142c89aa60d0fb2442863", id)
+	assert.NoError(t, err)
```
Comment (lukaszgryglicki) [inline utils/uuid/uuid_test.go:17]: If you write tests for UUID, please cover all edge cases and compare them with python results, not with go code. So keep few basic tests like this, but then test edge case results - your go code vs. uuid.py results - this will really tell us that we can use go implementation instead of python one - I remember that our current UUID implementations in go were giving a bit different results that python ones - this is why I've added an option to support so-called legacy-UUID.

**Review by lukaszgryglicki — CHANGES_REQUESTED** [review]
Please see my comments regarding UUID support. You may also take a look at Request() implementation in utils.go - it supports everything you implemented for HttpClientProvider plus cookies, retrying and optionally caching requests.

**Comment by lukaszgryglicki** [comment]
Also there is a conflict.

**Comment by lukaszgryglicki** [comment]
Also please note that `make` stopped working, please also address lint suggestions etc. Basically `make` should work. I know that this is a WIP PR but this is just information:
`make`:
```
./for_each_go_file.sh "gofmt -s -w"
./for_each_go_file.sh "golint -set_exit_status"
./utils/uuid/uuid.go:10:1: exported function Generate should have comment or be unexported
Found 1 lint suggestions; failing.
Makefile:32: recipe for target 'lint' failed
make: *** [lint] Error 1
```

File: cmd/dads/dads.go (hunk 1 above)
Comment (enkhalifapro) [reply cmd/dads/dads.go:20]: It's ready but regarding this part of ads running, we should rework it

File: utils/uuid/uuid.go (hunk 3 above)
Comment (enkhalifapro) [reply utils/uuid/uuid.go:2]: Yes, I rewrote it in go and added a unit test to make sure it's exactly the same as python!

File: utils/uuid/uuid_test.go (hunk 4 above)
Comment (enkhalifapro) [reply utils/uuid/uuid_test.go:17]: Ok!

File: utils/uuid/uuid.go (hunk 3 above)
Comment (lukaszgryglicki) [reply utils/uuid/uuid.go:2]: is it checking all edge cases  - especially empty, nil, UTF-8 values - see all specific python and go code and test all possible edge cases. If so then OK.

File: dockerhub/fetcher.go (hunk 2 above)
Comment (enkhalifapro) [reply dockerhub/fetcher.go:38]: Request has Ctx on it will will the entire coupled stuff into fetcher which will prevent testability, so I can't use it and I have to use a light and loosely coupled dependency

File: utils/uuid/uuid.go (hunk 3 above)
Comment (enkhalifapro) [reply utils/uuid/uuid.go:2]: Done and tested including mention edge cases, and moved to `LF-Engineering/dev-analytics-libraries`

**Comment by lukaszgryglicki** [comment]
Is this ready for review? If so please remove the WIP label...

**Comment by lukaszgryglicki** [comment]
Please sync with master and let me know when ready - it's a huge PR, so I won't review it today but can do it tomorrow.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Check https://circleci.com/docs/2.0/language-go/ for more details
+version: 2
+jobs:
+  build:
+    docker:
+      - image: golang:1.13
```
Comment (lukaszgryglicki) [inline .circleci/config.yml]: Please update to v1.15

File: Makefile [hunk 6]
```diff
@@ -20,23 +20,24 @@ GO_ERRCHECK=errcheck -asserts -ignore '[FS]?[Pp]rint*' -ignoretests
...
 	 ${GO_ENV} ${GO_BUILD} -o dads cmd/dads/dads.go
 
 fmt: ${GO_BIN_FILES} ${GO_LIB_FILES} ${GO_TEST_FILES} ${GO_LIBTEST_FILES}
-	./for_each_go_file.sh "${GO_FMT}"
+	./scripts/for_each_go_file.sh "${GO_FMT}"
 
-lint: ${GO_BIN_FILES} ${GO_LIB_FILES} ${GO_TEST_FILES} ${GO_LIBTEST_FILES}
-	./for_each_go_file.sh "${GO_LINT}"
+lint: ## Lint the files
+	golint -set_exit_status $(PKG_LIST)
```
Comment (lukaszgryglicki) [inline Makefile:34]: I was ignoring `[pP]rintf*` functions result on purpose - we don't need to handle their errors, not sure if that change will keep this or not, otherwise OK with this.

File: cmd/dads/dads.go [hunk 7]
```diff
@@ -1,7 +1,9 @@
 package main
 
 import (
+	"encoding/json"
```
Comment (lukaszgryglicki) [inline cmd/dads/dads.go:4]: I'm using jsoniter - its many times faster than a normal `json` - it has the same methods, just replace `json` with `jsoniter`.

File: dockerhub/fetcher_test.go [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    "is_private": false,
+    "is_automated": true,
+    "can_edit": false,
+    "star_count": 1,
+    "pull_count": 398,
+    "last_updated": "2017-05-10T08:12:52.217787Z",
+    "build_on_cloud": null,
+    "has_starred": false,
+    "full_description": "# Perceval [![Build Status](https://travis-ci.org/grimoirelab/perceval.svg?branch=master)](https://travis-ci.org/grimoirelab/perceval) [![Coverage Status](https://img.shields.io/coveralls/grimoirelab/perceval.svg)](https://coveralls.io/r/grimoirelab/perceval?branch=master)\n\nSend Sir Perceval on a quest to retrieve and gather data from software\nrepositories.\n\n## Usage\n\n\nusage:",
+"affiliation": null,
```
Comment (lukaszgryglicki) [inline dockerhub/fetcher_test.go:95]: whitespace/indenting

File: go.mod [hunk 9]
```diff
@@ -0,0 +1,16 @@
+module github.com/LF-Engineering/da-ds
+
+go 1.13
```
Comment (lukaszgryglicki) [inline go.mod]: Please update to `v1.15` as requested by me & @fayazg

**Review by lukaszgryglicki — CHANGES_REQUESTED** [review]
Few small changes... otherwise OK.

File: .circleci/config.yml (hunk 5 above)
Comment (enkhalifapro) [reply .circleci/config.yml]: Done

File: cmd/dads/dads.go (hunk 7 above)
Comment (lukaszgryglicki) [reply cmd/dads/dads.go:4]: can you change to `jsoniter`?

**Comment by lukaszgryglicki** [comment]
When I `git pull` your branch, `make` is not working:
```
2020-12-02 06:58:14 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# make
./scripts/for_each_go_file.sh "gofmt -s -w"
golint -set_exit_status github.com/LF-Engineering/da-ds github.com/LF-Engineering/da-ds/cmd/dads github.com/LF-Engineering/da-ds/dockerhub github.com/LF-Engineering/da-ds/dockerhub/mocks github.com/LF-Engineering/da-ds/test github.com/LF-Engineering/da-ds/utils
open github.com/LF-Engineering/da-ds: no such file or directory
open github.com/LF-Engineering/da-ds/cmd/dads: no such file or directory
open github.com/LF-Engineering/da-ds/dockerhub: no such file or directory
open github.com/LF-Engineering/da-ds/dockerhub/mocks: no such file or directory
open github.com/LF-Engineering/da-ds/test: no such file or directory
open github.com/LF-Engineering/da-ds/utils: no such file or directory
./scripts/for_each_go_file.sh "goimports -w"
./scripts/vet_files.sh "go vet"
/bin/sh: 1: ./scripts/vet_files.sh: not found
Makefile:37: recipe for target 'vet' failed
make: *** [vet] Error 127
```
I'm in the correct directory as you can see from the first line `~/dev/go/src/github.com/LF-Engineering/da-ds`
When I `git checkout` the master branch it complains about make results not saved (
Screenshot 2020-12-02 at 08 02 52
Screenshot 2020-12-02 at 08 03 02

This is because `gofmt` changes are not applied in your code.
Finally when I `git reset --hard` to master and do `make` it works correctly:
```
2020-12-02 07:03:02 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git reset --hard
HEAD is now at 3a8eb9d DA-2812: upgrade go version to 1.15
2020-12-02 07:04:02 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
2020-12-02 07:04:06 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# make
./for_each_go_file.sh "gofmt -s -w"
./for_each_go_file.sh "golint -set_exit_status"
./for_each_go_file.sh "goimports -w"
./vet_files.sh "go vet"
goconst ./...
usedexports ./...
errcheck -asserts -ignore '[FS]?[Pp]rint*' -ignoretests ./...
CGO_ENABLED=0 go build -ldflags '-s -w' -o dads cmd/dads/dads.go
2020-12-02 07:04:16 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# ls -l dads
-rwxr-xr-x 1 root root 11464704 Dec  2 07:04 dads
2020-12-02 07:04:23 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# 
```

File: cmd/dads/dads.go (hunk 7 above)
Comment (enkhalifapro) [reply cmd/dads/dads.go]: Done

File: dockerhub/fetcher_test.go (hunk 8 above)
Comment (enkhalifapro) [reply dockerhub/fetcher_test.go:96]: Done!

File: go.mod (hunk 9 above)
Comment (enkhalifapro) [reply go.mod]: Done!

File: uuid.go [hunk 10]
```diff
@@ -107,7 +109,7 @@ func UUIDAffs(ctx *Ctx, args ...string) (h string) {
 	}
 	var err error
 	if len(args) != 4 {
-		err = fmt.Errorf("GenerateIdentity requires exactly 4 asrguments, got %+v\n", args)
+		err = fmt.Errorf("generateIdentity requires exactly 4 asrguments, got %+v", args)
```
Comment (lukaszgryglicki) [inline uuid.go:112]: Actually there is a typo here: `asrguments` - but not requesting to fix it now.

File: uuid.go [hunk 11]
```diff
@@ -56,7 +56,9 @@ func UUIDNonEmpty(ctx *Ctx, args ...string) (h string) {
 		cmdLine := []string{"uuid.py", "a"}
 		cmdLine = append(cmdLine, args...)
 		h, _, err = ExecCommand(ctx, cmdLine, "", nil)
-		FatalOnError(err)
+		if err != nil {
+			return ""
+		}
 		h = h[:len(h)-1]
```
Comment (lukaszgryglicki) [inline uuid.go:62]: So we ignore the error (I don't think there should be an error in UUID generation) and return an empty UUID instead? That will be wrong, UUID must be non-empty and fixed length.

**Comment by lukaszgryglicki** [comment]
Still when I `git pull` this branch and then execute make I get (see the full workflow):
```
2020-12-02 16:36:52 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git checkout DA-2812-write-dockerhub-instrumentation-in-golang
Switched to branch 'DA-2812-write-dockerhub-instrumentation-in-golang'
Your branch is up to date with 'origin/DA-2812-write-dockerhub-instrumentation-in-golang'.
2020-12-02 16:36:55 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git pull
remote: Enumerating objects: 50, done.
remote: Counting objects: 100% (43/43), done.
remote: Compressing objects: 100% (11/11), done.
remote: Total 28 (delta 19), reused 22 (delta 13), pack-reused 0
Unpacking objects: 100% (28/28), done.
From https://github.com/LF-Engineering/da-ds
   3a8eb9d..9cdb6b7  DA-2812-write-dockerhub-instrumentation-in-golang -> origin/DA-2812-write-dockerhub-instrumentation-in-golang
   25963c4..cdf6a11  DA-2812-unit-testing                              -> origin/DA-2812-unit-testing
Updating 3a8eb9d..9cdb6b7
Fast-forward
 cmd/dads/dads.go           |   9 ++--
 dockerhub/enricher.go      |   6 ++-
 dockerhub/enricher_test.go | 680 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------------------------------------------------------------------------------
 dockerhub/fetcher.go       |  21 +++++----
 dockerhub/fetcher_test.go  | 181 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----
 dockerhub/manager_test.go  |  28 -----------
 json.go                    |   6 +--
 utils/es_client.go         |  10 ++--
 8 files changed, 654 insertions(+), 287 deletions(-)
 delete mode 100644 dockerhub/manager_test.go
2020-12-02 16:37:01 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# make
./scripts/for_each_go_file.sh "gofmt -s -w"
golint -set_exit_status github.com/LF-Engineering/da-ds github.com/LF-Engineering/da-ds/cmd/dads github.com/LF-Engineering/da-ds/dockerhub github.com/LF-Engineering/da-ds/dockerhub/mocks github.com/LF-Engineering/da-ds/test github.com/LF-Engineering/da-ds/utils
open github.com/LF-Engineering/da-ds: no such file or directory
open github.com/LF-Engineering/da-ds/cmd/dads: no such file or directory
open github.com/LF-Engineering/da-ds/dockerhub: no such file or directory
open github.com/LF-Engineering/da-ds/dockerhub/mocks: no such file or directory
open github.com/LF-Engineering/da-ds/test: no such file or directory
open github.com/LF-Engineering/da-ds/utils: no such file or directory
./scripts/for_each_go_file.sh "goimports -w"
./scripts/vet_files.sh "go vet"
/bin/sh: 1: ./scripts/vet_files.sh: not found
Makefile:37: recipe for target 'vet' failed
make: *** [vet] Error 127
2020-12-02 16:37:09 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git status
On branch DA-2812-write-dockerhub-instrumentation-in-golang
Your branch is up to date with 'origin/DA-2812-write-dockerhub-instrumentation-in-golang'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   cmd/dads/dads.go
	modified:   dockerhub/const.go
	modified:   dockerhub/enricher.go
	modified:   dockerhub/manager.go
	modified:   dockerhub/mocks/es_client_provider.go
	modified:   utils/es_client.go
	modified:   utils/http_client.go
	modified:   utils/time.go
	modified:   utils/time_test.go

no changes added to commit (use "git add" and/or "git commit -a")
2020-12-02 16:37:33 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git checkout master
error: Your local changes to the following files would be overwritten by checkout:
	cmd/dads/dads.go
	dockerhub/const.go
	dockerhub/enricher.go
	dockerhub/manager.go
	dockerhub/mocks/es_client_provider.go
	utils/es_client.go
	utils/http_client.go
	utils/time.go
	utils/time_test.go
Please commit your changes or stash them before you switch branches.
Aborting
2020-12-02 16:38:12 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git reset --hard
HEAD is now at 9cdb6b7 DA-2812: code clean up
2020-12-02 16:38:16 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
2020-12-02 16:38:21 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# make
./for_each_go_file.sh "gofmt -s -w"
./for_each_go_file.sh "golint -set_exit_status"
./for_each_go_file.sh "goimports -w"
./vet_files.sh "go vet"
goconst ./...
usedexports ./...
errcheck -asserts -ignore '[FS]?[Pp]rint*' -ignoretests ./...
CGO_ENABLED=0 go build -ldflags '-s -w' -o dads cmd/dads/dads.go
2020-12-02 16:38:34 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# 
```

cc @1010sachin @enkhalifapro 

Please fix this before merging.

File: uuid.go (hunk 11 above)
Comment (enkhalifapro) [reply uuid.go:62]: This one should no longer used, all code should reference uuid module of libraries repo

Comment (lukaszgryglicki) [reply uuid.go:62]: No, this `if branch` *is* and *will be* used when legacy mode *is* selected. Your branch (UUID from libraries is used by default in both affs UUID and document UUID).

**Comment by lukaszgryglicki** [comment]
@enkhalifapro really *almost*
`make build` works fine but `make` fails, only because `BINARIES` definition in `Makefile` changed, see output of `make build` and `make`:
```
2020-12-03 14:42:53 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# make clean
rm -f dads
2020-12-03 14:42:56 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# make
./scripts/for_each_go_file.sh "gofmt -s -w"
golint -set_exit_status github.com/LF-Engineering/da-ds github.com/LF-Engineering/da-ds/cmd/dads github.com/LF-Engineering/da-ds/dockerhub github.com/LF-Engineering/da-ds/test github.com/LF-Engineering/da-ds/utils
./scripts/for_each_go_file.sh "goimports -w"
go vet github.com/LF-Engineering/da-ds github.com/LF-Engineering/da-ds/cmd/dads github.com/LF-Engineering/da-ds/dockerhub github.com/LF-Engineering/da-ds/test github.com/LF-Engineering/da-ds/utils
usedexports ./...
errcheck -verbose -asserts -ignoretests -ignoregenerated ./...
Excluding (*bytes.Buffer).Write
Excluding (*bytes.Buffer).WriteByte
Excluding (*bytes.Buffer).WriteRune
Excluding (*bytes.Buffer).WriteString
Excluding fmt.Errorf
Excluding fmt.Print
Excluding fmt.Printf
Excluding fmt.Println
Excluding fmt.Fprint(*bytes.Buffer)
Excluding fmt.Fprintf(*bytes.Buffer)
Excluding fmt.Fprintln(*bytes.Buffer)
Excluding fmt.Fprint(*strings.Builder)
Excluding fmt.Fprintf(*strings.Builder)
Excluding fmt.Fprintln(*strings.Builder)
Excluding fmt.Fprint(os.Stderr)
Excluding fmt.Fprintf(os.Stderr)
Excluding fmt.Fprintln(os.Stderr)
Excluding math/rand.Read
Excluding (*math/rand.Rand).Read
Excluding (*strings.Builder).Write
Excluding (*strings.Builder).WriteByte
Excluding (*strings.Builder).WriteRune
Excluding (*strings.Builder).WriteString
Excluding (hash.Hash).Write
Checking github.com/LF-Engineering/da-ds
Checking github.com/LF-Engineering/da-ds/utils
Checking github.com/LF-Engineering/da-ds/dockerhub
Checking github.com/LF-Engineering/da-ds/cmd/dads
Checking github.com/LF-Engineering/da-ds/dockerhub/mocks
Checking github.com/LF-Engineering/da-ds/test
make: *** No rule to make target 'dads', needed by 'all'.  Stop.
2020-12-03 14:43:04 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# make build
CGO_ENABLED=0 go build -ldflags '-s -w' -o dads cmd/dads/dads.go
2020-12-03 14:43:17 root@teststats:~/dev/go/src/github.com/LF-Engineering/da-ds# 
```

**Comment by lukaszgryglicki** [comment]
Note simple error: `make: *** No rule to make target 'dads', needed by 'all'.  Stop.`, while`make builds` works.
